// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  ScanbotBarcodeCameraView: true,
  Point: true
};
Object.defineProperty(exports, "Point", {
  enumerable: true,
  get: function () {
    return _utils.Point;
  }
});
Object.defineProperty(exports, "ScanbotBarcodeCameraView", {
  enumerable: true,
  get: function () {
    return _ScanbotBarcodeCameraView.ScanbotBarcodeCameraView;
  }
});
exports.default = void 0;
var _reactNative = require("react-native");
var _ScanbotBarcodeCameraView = require("./component/barcode-camera-view/ScanbotBarcodeCameraView");
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
var _customTypes = require("./customTypes");
Object.keys(_customTypes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _customTypes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _customTypes[key];
    }
  });
});
var _results = require("./results");
Object.keys(_results).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _results[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _results[key];
    }
  });
});
var _configurations = require("./configurations");
Object.keys(_configurations).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _configurations[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _configurations[key];
    }
  });
});
var _customConfigurations = require("./customConfigurations");
Object.keys(_customConfigurations).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _customConfigurations[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _customConfigurations[key];
    }
  });
});
var _utils = require("./utils");
var _BarcodeDocumentModel = require("./documents/BarcodeDocumentModel");
Object.keys(_BarcodeDocumentModel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _BarcodeDocumentModel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _BarcodeDocumentModel[key];
    }
  });
});
var _CommonFieldType = require("./documents/CommonFieldType");
Object.keys(_CommonFieldType).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _CommonFieldType[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _CommonFieldType[key];
    }
  });
});
var _GenericDocument = require("./documents/GenericDocument");
Object.keys(_GenericDocument).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _GenericDocument[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _GenericDocument[key];
    }
  });
});
var _ScanbotBarcodeCameraViewProperties = require("./component/barcode-camera-view/ScanbotBarcodeCameraViewProperties");
Object.keys(_ScanbotBarcodeCameraViewProperties).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ScanbotBarcodeCameraViewProperties[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ScanbotBarcodeCameraViewProperties[key];
    }
  });
});
const LINKING_ERROR = `The package 'react-native-scanbot-barcode-scanner-sdk' doesn't seem to be linked. Make sure: \n\n` + _reactNative.Platform.select({
  ios: "- You have run 'pod install'\n",
  default: ''
}) + '- You rebuilt the app after installing the package\n' + '- You are not using Expo Go\n';
const ScanbotBarcodeSDKImpl = _reactNative.NativeModules.ScanbotBarcodeSdk ? _reactNative.NativeModules.ScanbotBarcodeSdk : new Proxy({}, {
  get() {
    throw new Error(LINKING_ERROR);
  }
});
const ScanbotBarcodeSDK = {
  /**
   * Initialize the Scanbot Barcode Scanner SDK with the preferred configuration.
   *
   * @param {ScanbotBarcodeSdkConfiguration} config
   * @returns {Promise<ResultWrapper<string>>}
   */
  initializeSdk: config => {
    return ScanbotBarcodeSDKImpl.initializeSdk(config);
  },
  /**
   * Provides complete information about the current license status.
   *
   * @returns {Promise<ResultWrapper<LicenseInfoResult>>}
   */
  getLicenseInfo: () => {
    return ScanbotBarcodeSDKImpl.getLicenseInfo();
  },
  /**
   * Remove all files generated by this plugin.
   *
   * @returns {Promise<ResultWrapper<string>>}
   */
  cleanup: () => {
    return ScanbotBarcodeSDKImpl.cleanup();
  },
  /**
   * Opens the Ready-To-Use UI screen for barcode scanning with the desired configuration.
   *
   * @param {BarcodeScannerConfiguration} config
   * @returns {Promise<ResultWrapper<BarcodeScannerResult>>}
   *
   * @deprecated Use ***startBarcodeScanner*** from ***'react-native-scanbot-barcode-scanner-sdk/ui_v2'*** instead.
   */
  startBarcodeScanner: config => {
    return ScanbotBarcodeSDKImpl.startBarcodeScanner(config);
  },
  /**
   * Force the barcode scanning Ready-To-Use UI screen to close while it is running.
   *
   * @returns {Promise<ResultWrapper<undefined>>}
   *
   * @deprecated
   */
  closeBarcodeScanner: () => {
    return ScanbotBarcodeSDKImpl.closeBarcodeScanner();
  },
  /**
   * Opens the Ready-To-Use UI screen for batch barcode scanning with the desired configuration.
   *
   * @param {BatchBarcodeScannerConfiguration} config
   * @returns {Promise<ResultWrapper<BatchBarcodeScannerResult>>}
   *
   * @deprecated Use ***startBarcodeScanner*** from ***'react-native-scanbot-barcode-scanner-sdk/ui_v2'*** instead.
   */
  startBatchBarcodeScanner: config => {
    return ScanbotBarcodeSDKImpl.startBatchBarcodeScanner(config);
  },
  /**
   * Force the batch barcode scanning Ready-To-Use UI screen to close while it is running.
   *
   * @returns {Promise<ResultWrapper<undefined>>}
   *
   * @deprecated
   */
  closeBatchBarcodeScanner: () => {
    return ScanbotBarcodeSDKImpl.closeBatchBarcodeScanner();
  },
  /**
   * Detect barcodes on the image represented by the file URL. The image file URL is part of the input arguments.
   *
   * @param {DetectBarcodesOnImageArguments} args
   * @returns {Promise<ResultWrapper<BarcodeScannerResult>>}
   */
  detectBarcodesOnImage: args => {
    return ScanbotBarcodeSDKImpl.detectBarcodesOnImage(args);
  },
  /**
   * Extract images from a PDF represented by the file URL. The PDF file URL is part of the input arguments.
   *
   * @param {ExtractImagesFromPdfArguments} args
   * @returns {Promise<ResultWrapper<string[]>>}
   */
  extractImagesFromPDF: args => {
    return ScanbotBarcodeSDKImpl.extractImagesFromPDF(args);
  },
  /**
   * Returns the Base 64 encoded representation of the image data.
   */
  getImageData(imageFileUri) {
    return ScanbotBarcodeSDKImpl.getImageData(imageFileUri);
  }
};
var _default = exports.default = ScanbotBarcodeSDK;
