// react-native-scanbot-barcode-scanner-sdk/src/ui_v2/barcode/BarcodeTextLocalization.ts
import { DeepPartial, PartiallyConstructible } from '../../utils';

/**
Configuration of the all strings for barcode scanner.
*/
export class BarcodeTextLocalization extends PartiallyConstructible {
  /**
    Title for the camera permission dialog.
    Default is "Camera permission denied!"
    */
  public cameraPermissionEnableCameraTitle: string = 'Camera permission denied!';
  /**
    Explanation for the camera permission dialog.
    Default is "Please allow the usage of the camera to start the scanning process."
    */
  public cameraPermissionEnableCameraExplanation: string =
    'Please allow the usage of the camera to start the scanning process.';
  /**
    Button text for the camera permission dialog.
    Default is "Grant permission"
    */
  public cameraPermissionEnableCameraButton: string = 'Grant permission';
  /**
    Close button text for the camera permission dialog.
    Default is "Close"
    */
  public cameraPermissionCloseButton: string = 'Close';
  /**
    Cancel button text for the top bar.
    Default is "Cancel"
    */
  public topBarCancelButton: string = 'Cancel';
  /**
    Title for the top bar.
    Default is "Scan Item"
    */
  public topBarTitle: string = 'Scan Item';
  /**
    User guidance text.
    Default is "Move the finder over a barcode"
    */
  public userGuidance: string = 'Move the finder over a barcode';
  /**
    Loading message for barcode info mapping.
    Default is "Loading details..."
    */
  public barcodeInfoMappingLoadingMessage: string = 'Loading details...';
  /**
    Title for the barcode info mapping error state.
    Default is "Connection Error!"
    */
  public barcodeInfoMappingErrorStateTitle: string = 'Connection Error!';
  /**
    Subtitle for the barcode info mapping error state.
    Default is "There was an issue and the data requested was not fetched. You could try again or discard this result to start a new scan."
    */
  public barcodeInfoMappingErrorStateSubtitle: string =
    'There was an issue and the data requested was not fetched. You could try again or discard this result to start a new scan.';
  /**
    Retry button text for the barcode info mapping error state.
    Default is "Retry"
    */
  public barcodeInfoMappingErrorStateRetryButton: string = 'Retry';
  /**
    Cancel button text for the barcode info mapping error state.
    Default is "Cancel"
    */
  public barcodeInfoMappingErrorStateCancelButton: string = 'Cancel';
  /**
    Title for the manual count edit dialog.
    Default is "Update count"
    */
  public manualCountEditDialogTitle: string = 'Update count';
  /**
    Info for the manual count edit dialog.
    Default is "Adjust the number of items you scanned."
    */
  public manualCountEditDialogInfo: string = 'Adjust the number of items you scanned.';
  /**
    Update button text for the manual count edit dialog.
    Default is "Update"
    */
  public manualCountEditDialogUpdateButton: string = 'Update';
  /**
    Cancel button text for the manual count edit dialog.
    Default is "Cancel"
    */
  public manualCountEditDialogCancelButton: string = 'Cancel';
  /**
    Reset button text for the barcode sheet.
    Default is "Reset"
    */
  public sheetResetButton: string = 'Reset';
  /**
    Clear all button text for the barcode sheet.
    Default is "Clear all"
    */
  public sheetClearAllButton: string = 'Clear all';
  /**
    Submit button text for the barcode sheet.
    Default is "Submit"
    */
  public sheetSubmitButton: string = 'Submit';
  /**
    Start scanning button text for the barcode sheet.
    Default is "Start scanning"
    */
  public sheetStartScanningButton: string = 'Start scanning';
  /**
    Title for the barcode sheet placeholder.
    Default is "No barcodes here!"
    */
  public sheetPlaceholderTitle: string = 'No barcodes here!';
  /**
    Subtitle for the barcode sheet placeholder.
    Default is "The barcode list is currently empty. Close this sheet and scan your items to add them."
    */
  public sheetPlaceholderSubtitle: string =
    'The barcode list is currently empty. Close this sheet and scan your items to add them.';
  /**
    Cancel button text for the single mode confirmation dialog.
    Default is "Close"
    */
  public singleModeConfirmationCancelButton: string = 'Close';
  /**
    Title for the multiple mode barcode sheet.
    Default is "%d items"
    */
  public multipleModeSheetTitle: string = '%d items';
  /**
    Title for the find and pick barcode sheet.
    Default is "%d/%d scanned"
    */
  public findAndPickSheetTitle: string = '%d/%d scanned';
  /**
    Subtitle for the find and pick barcode sheet.
    Default is "%d/%d added"
    */
  public findAndPickSheetBarcodeItemSubtitle: string = '%d/%d added';
  /**
    Title for the find and pick partial alert.
    Default is "Incomplete scan!"
    */
  public findAndPickPartialAlertTitle: string = 'Incomplete scan!';
  /**
    Subtitle for the find and pick partial alert.
    Default is "Some codes are yet to be scanned. Proceed with submission anyway?"
    */
  public findAndPickPartialAlertSubtitle: string =
    'Some codes are yet to be scanned. Proceed with submission anyway?';
  /**
    Cancel button text for the find and pick partial alert.
    Default is "Cancel"
    */
  public findAndPickPartialAlertCancelButton: string = 'Cancel';
  /**
    Submit button text for the find and pick partial alert.
    Default is "Yes, Submit"
    */
  public findAndPickPartialAlertSubmitButton: string = 'Yes, Submit';
  /**
    Title for the find and pick complete alert.
    Default is "All codes found!"
    */
  public findAndPickCompleteAlertTitle: string = 'All codes found!';
  /**
    Subtitle for the find and pick complete alert.
    Default is "You have successfully scanned all codes. Proceed with submission?"
    */
  public findAndPickCompleteAlertSubtitle: string =
    'You have successfully scanned all codes. Proceed with submission?';
  /**
    Cancel button text for the find and pick complete alert.
    Default is "Dismiss"
    */
  public findAndPickCompleteAlertCancelButton: string = 'Dismiss';
  /**
    Submit button text for the find and pick complete alert.
    Default is "Submit"
    */
  public findAndPickCompleteAlertSubmitButton: string = 'Submit';
  /**
    Accessibility description for the flash button.
    Default is "Toggle flash"
    */
  public accessibilityDescriptionFlashButton: string = 'Toggle flash';
  /**
    Accessibility description for the zoom button.
    Default is "Toggle camera zoom"
    */
  public accessibilityDescriptionZoomButton: string = 'Toggle camera zoom';
  /**
    Accessibility description for the flip camera button.
    Default is "Flip camera"
    */
  public accessibilityDescriptionFlipCameraButton: string = 'Flip camera';

  /** @param source {@displayType `DeepPartial<BarcodeTextLocalization>`} */
  public constructor(source: DeepPartial<BarcodeTextLocalization> = {}) {
    super();
    if (source.cameraPermissionEnableCameraTitle !== undefined) {
      this.cameraPermissionEnableCameraTitle = source.cameraPermissionEnableCameraTitle;
    }
    if (source.cameraPermissionEnableCameraExplanation !== undefined) {
      this.cameraPermissionEnableCameraExplanation = source.cameraPermissionEnableCameraExplanation;
    }
    if (source.cameraPermissionEnableCameraButton !== undefined) {
      this.cameraPermissionEnableCameraButton = source.cameraPermissionEnableCameraButton;
    }
    if (source.cameraPermissionCloseButton !== undefined) {
      this.cameraPermissionCloseButton = source.cameraPermissionCloseButton;
    }
    if (source.topBarCancelButton !== undefined) {
      this.topBarCancelButton = source.topBarCancelButton;
    }
    if (source.topBarTitle !== undefined) {
      this.topBarTitle = source.topBarTitle;
    }
    if (source.userGuidance !== undefined) {
      this.userGuidance = source.userGuidance;
    }
    if (source.barcodeInfoMappingLoadingMessage !== undefined) {
      this.barcodeInfoMappingLoadingMessage = source.barcodeInfoMappingLoadingMessage;
    }
    if (source.barcodeInfoMappingErrorStateTitle !== undefined) {
      this.barcodeInfoMappingErrorStateTitle = source.barcodeInfoMappingErrorStateTitle;
    }
    if (source.barcodeInfoMappingErrorStateSubtitle !== undefined) {
      this.barcodeInfoMappingErrorStateSubtitle = source.barcodeInfoMappingErrorStateSubtitle;
    }
    if (source.barcodeInfoMappingErrorStateRetryButton !== undefined) {
      this.barcodeInfoMappingErrorStateRetryButton = source.barcodeInfoMappingErrorStateRetryButton;
    }
    if (source.barcodeInfoMappingErrorStateCancelButton !== undefined) {
      this.barcodeInfoMappingErrorStateCancelButton =
        source.barcodeInfoMappingErrorStateCancelButton;
    }
    if (source.manualCountEditDialogTitle !== undefined) {
      this.manualCountEditDialogTitle = source.manualCountEditDialogTitle;
    }
    if (source.manualCountEditDialogInfo !== undefined) {
      this.manualCountEditDialogInfo = source.manualCountEditDialogInfo;
    }
    if (source.manualCountEditDialogUpdateButton !== undefined) {
      this.manualCountEditDialogUpdateButton = source.manualCountEditDialogUpdateButton;
    }
    if (source.manualCountEditDialogCancelButton !== undefined) {
      this.manualCountEditDialogCancelButton = source.manualCountEditDialogCancelButton;
    }
    if (source.sheetResetButton !== undefined) {
      this.sheetResetButton = source.sheetResetButton;
    }
    if (source.sheetClearAllButton !== undefined) {
      this.sheetClearAllButton = source.sheetClearAllButton;
    }
    if (source.sheetSubmitButton !== undefined) {
      this.sheetSubmitButton = source.sheetSubmitButton;
    }
    if (source.sheetStartScanningButton !== undefined) {
      this.sheetStartScanningButton = source.sheetStartScanningButton;
    }
    if (source.sheetPlaceholderTitle !== undefined) {
      this.sheetPlaceholderTitle = source.sheetPlaceholderTitle;
    }
    if (source.sheetPlaceholderSubtitle !== undefined) {
      this.sheetPlaceholderSubtitle = source.sheetPlaceholderSubtitle;
    }
    if (source.singleModeConfirmationCancelButton !== undefined) {
      this.singleModeConfirmationCancelButton = source.singleModeConfirmationCancelButton;
    }
    if (source.multipleModeSheetTitle !== undefined) {
      this.multipleModeSheetTitle = source.multipleModeSheetTitle;
    }
    if (source.findAndPickSheetTitle !== undefined) {
      this.findAndPickSheetTitle = source.findAndPickSheetTitle;
    }
    if (source.findAndPickSheetBarcodeItemSubtitle !== undefined) {
      this.findAndPickSheetBarcodeItemSubtitle = source.findAndPickSheetBarcodeItemSubtitle;
    }
    if (source.findAndPickPartialAlertTitle !== undefined) {
      this.findAndPickPartialAlertTitle = source.findAndPickPartialAlertTitle;
    }
    if (source.findAndPickPartialAlertSubtitle !== undefined) {
      this.findAndPickPartialAlertSubtitle = source.findAndPickPartialAlertSubtitle;
    }
    if (source.findAndPickPartialAlertCancelButton !== undefined) {
      this.findAndPickPartialAlertCancelButton = source.findAndPickPartialAlertCancelButton;
    }
    if (source.findAndPickPartialAlertSubmitButton !== undefined) {
      this.findAndPickPartialAlertSubmitButton = source.findAndPickPartialAlertSubmitButton;
    }
    if (source.findAndPickCompleteAlertTitle !== undefined) {
      this.findAndPickCompleteAlertTitle = source.findAndPickCompleteAlertTitle;
    }
    if (source.findAndPickCompleteAlertSubtitle !== undefined) {
      this.findAndPickCompleteAlertSubtitle = source.findAndPickCompleteAlertSubtitle;
    }
    if (source.findAndPickCompleteAlertCancelButton !== undefined) {
      this.findAndPickCompleteAlertCancelButton = source.findAndPickCompleteAlertCancelButton;
    }
    if (source.findAndPickCompleteAlertSubmitButton !== undefined) {
      this.findAndPickCompleteAlertSubmitButton = source.findAndPickCompleteAlertSubmitButton;
    }
    if (source.accessibilityDescriptionFlashButton !== undefined) {
      this.accessibilityDescriptionFlashButton = source.accessibilityDescriptionFlashButton;
    }
    if (source.accessibilityDescriptionZoomButton !== undefined) {
      this.accessibilityDescriptionZoomButton = source.accessibilityDescriptionZoomButton;
    }
    if (source.accessibilityDescriptionFlipCameraButton !== undefined) {
      this.accessibilityDescriptionFlipCameraButton =
        source.accessibilityDescriptionFlipCameraButton;
    }
  }
}
