// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/ui_v2/common/ScanbotAlertDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScanbotAlertDialog = void 0;
var _Common = require("../common/Common");
var _utils = require("../../utils");
/**
Configuration of the the standard alert dialog.
*/
class ScanbotAlertDialog extends _utils.PartiallyConstructible {
  /**
    The title displayed above the message.
    */
  title = new _Common.StyledText({
    text: 'Title',
    color: '?sbColorOnSurface'
  });
  /**
    The explanation message.
    */
  subtitle = new _Common.StyledText({
    text: 'Standard explanation message text.',
    color: '?sbColorOnSurfaceVariant'
  });
  /**
    The background color of the alert dialog.
    Default is "?sbColorSurface"
    */
  sheetColor = '?sbColorSurface';
  /**
    The dialog overlay color.
    Default is "?sbColorModalOverlay"
    */
  modalOverlayColor = '?sbColorModalOverlay';
  /**
    The color of the divider line.
    Default is "?sbColorOutline"
    */
  dividerColor = '?sbColorOutline';
  /**
    Configuration of the 'OK' button.
    */
  okButton = new _Common.ButtonConfiguration({
    background: new _Common.BackgroundStyle({
      fillColor: '?sbColorPrimary'
    })
  });
  /**
    Configuration of the middle button for dialogs with tree buttons.
    */
  actionButton = new _Common.ButtonConfiguration({
    visible: false,
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '#00000000',
      strokeWidth: 0.0
    })
  });
  /**
    Configuration of the 'cancel' button.
    */
  cancelButton = new _Common.ButtonConfiguration({
    visible: true,
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '#00000000',
      strokeWidth: 0.0
    })
  });

  /** @param source {@displayType `DeepPartial<ScanbotAlertDialog>`} */
  constructor(source = {}) {
    super();
    if (source.title !== undefined) {
      this.title = new _Common.StyledText(source.title);
    }
    if (source.subtitle !== undefined) {
      this.subtitle = new _Common.StyledText(source.subtitle);
    }
    if (source.sheetColor !== undefined) {
      this.sheetColor = source.sheetColor;
    }
    if (source.modalOverlayColor !== undefined) {
      this.modalOverlayColor = source.modalOverlayColor;
    }
    if (source.dividerColor !== undefined) {
      this.dividerColor = source.dividerColor;
    }
    if (source.okButton !== undefined) {
      this.okButton = new _Common.ButtonConfiguration(source.okButton);
    }
    if (source.actionButton !== undefined) {
      this.actionButton = new _Common.ButtonConfiguration(source.actionButton);
    }
    if (source.cancelButton !== undefined) {
      this.cancelButton = new _Common.ButtonConfiguration(source.cancelButton);
    }
  }
}
exports.ScanbotAlertDialog = ScanbotAlertDialog;
