// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/ui_v2/common/UserGuidanceConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserGuidanceConfiguration = exports.IconUserGuidanceConfiguration = void 0;
var _Common = require("../common/Common");
var _utils = require("../../utils");
/**
Configuration of the hint guiding users through the scanning process.
*/
class UserGuidanceConfiguration extends _utils.PartiallyConstructible {
  /**
    Determines whether the user guidance is visible or not.
    Default is true
    */
  visible = true;
  /**
    The title of the user guidance.
    */
  title = new _Common.StyledText({
    color: '?sbColorOnPrimary'
  });
  /**
    The background style used for the user guidance.
    */
  background = new _Common.BackgroundStyle({
    strokeColor: '#00000000',
    fillColor: '?sbColorSurfaceLow'
  });

  /** @param source {@displayType `DeepPartial<UserGuidanceConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.title !== undefined) {
      this.title = new _Common.StyledText(source.title);
    }
    if (source.background !== undefined) {
      this.background = new _Common.BackgroundStyle(source.background);
    }
  }
}

/**
Configuration of the hint (containing an icon) guiding users through the scanning process.
*/
exports.UserGuidanceConfiguration = UserGuidanceConfiguration;
class IconUserGuidanceConfiguration extends _utils.PartiallyConstructible {
  /**
    Determines whether the user guidance is visible or not.
    Default is true
    */
  visible = true;
  /**
    Configuration of the icon appearance.
    */
  icon = new _Common.IconStyle({
    color: '?sbColorOnPrimary'
  });
  /**
    The title of the user guidance.
    */
  title = new _Common.StyledText({
    color: '?sbColorOnPrimary'
  });
  /**
    Configuration of the background appearance for the user guidance hints.
    */
  background = new _Common.BackgroundStyle({
    fillColor: '?sbColorSurfaceLow'
  });

  /** @param source {@displayType `DeepPartial<IconUserGuidanceConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.icon !== undefined) {
      this.icon = new _Common.IconStyle(source.icon);
    }
    if (source.title !== undefined) {
      this.title = new _Common.StyledText(source.title);
    }
    if (source.background !== undefined) {
      this.background = new _Common.BackgroundStyle(source.background);
    }
  }
}
exports.IconUserGuidanceConfiguration = IconUserGuidanceConfiguration;
