// react-native-scanbot-barcode-scanner-sdk/lib/module/ui_v2/barcode/BarcodeTextLocalization.js
import { PartiallyConstructible } from '../../utils';

/**
Configuration of the all strings for barcode scanner.
*/
export class BarcodeTextLocalization extends PartiallyConstructible {
  /**
    Title for the camera permission dialog.
    Default is "Camera permission denied!"
    */
  cameraPermissionEnableCameraTitle = 'Camera permission denied!';
  /**
    Explanation for the camera permission dialog.
    Default is "Please allow the usage of the camera to start the scanning process."
    */
  cameraPermissionEnableCameraExplanation = 'Please allow the usage of the camera to start the scanning process.';
  /**
    Button text for the camera permission dialog.
    Default is "Grant permission"
    */
  cameraPermissionEnableCameraButton = 'Grant permission';
  /**
    Close button text for the camera permission dialog.
    Default is "Close"
    */
  cameraPermissionCloseButton = 'Close';
  /**
    Cancel button text for the top bar.
    Default is "Cancel"
    */
  topBarCancelButton = 'Cancel';
  /**
    Title for the top bar.
    Default is "Scan Item"
    */
  topBarTitle = 'Scan Item';
  /**
    User guidance text.
    Default is "Move the finder over a barcode"
    */
  userGuidance = 'Move the finder over a barcode';
  /**
    Loading message for barcode info mapping.
    Default is "Loading details..."
    */
  barcodeInfoMappingLoadingMessage = 'Loading details...';
  /**
    Title for the barcode info mapping error state.
    Default is "Connection Error!"
    */
  barcodeInfoMappingErrorStateTitle = 'Connection Error!';
  /**
    Subtitle for the barcode info mapping error state.
    Default is "There was an issue and the data requested was not fetched. You could try again or discard this result to start a new scan."
    */
  barcodeInfoMappingErrorStateSubtitle = 'There was an issue and the data requested was not fetched. You could try again or discard this result to start a new scan.';
  /**
    Retry button text for the barcode info mapping error state.
    Default is "Retry"
    */
  barcodeInfoMappingErrorStateRetryButton = 'Retry';
  /**
    Cancel button text for the barcode info mapping error state.
    Default is "Cancel"
    */
  barcodeInfoMappingErrorStateCancelButton = 'Cancel';
  /**
    Title for the manual count edit dialog.
    Default is "Update count"
    */
  manualCountEditDialogTitle = 'Update count';
  /**
    Info for the manual count edit dialog.
    Default is "Adjust the number of items you scanned."
    */
  manualCountEditDialogInfo = 'Adjust the number of items you scanned.';
  /**
    Update button text for the manual count edit dialog.
    Default is "Update"
    */
  manualCountEditDialogUpdateButton = 'Update';
  /**
    Cancel button text for the manual count edit dialog.
    Default is "Cancel"
    */
  manualCountEditDialogCancelButton = 'Cancel';
  /**
    Reset button text for the barcode sheet.
    Default is "Reset"
    */
  sheetResetButton = 'Reset';
  /**
    Clear all button text for the barcode sheet.
    Default is "Clear all"
    */
  sheetClearAllButton = 'Clear all';
  /**
    Submit button text for the barcode sheet.
    Default is "Submit"
    */
  sheetSubmitButton = 'Submit';
  /**
    Start scanning button text for the barcode sheet.
    Default is "Start scanning"
    */
  sheetStartScanningButton = 'Start scanning';
  /**
    Title for the barcode sheet placeholder.
    Default is "No barcodes here!"
    */
  sheetPlaceholderTitle = 'No barcodes here!';
  /**
    Subtitle for the barcode sheet placeholder.
    Default is "The barcode list is currently empty. Close this sheet and scan your items to add them."
    */
  sheetPlaceholderSubtitle = 'The barcode list is currently empty. Close this sheet and scan your items to add them.';
  /**
    Cancel button text for the single mode confirmation dialog.
    Default is "Close"
    */
  singleModeConfirmationCancelButton = 'Close';
  /**
    Title for the multiple mode barcode sheet.
    Default is "%d items"
    */
  multipleModeSheetTitle = '%d items';
  /**
    Title for the find and pick barcode sheet.
    Default is "%d/%d scanned"
    */
  findAndPickSheetTitle = '%d/%d scanned';
  /**
    Subtitle for the find and pick barcode sheet.
    Default is "%d/%d added"
    */
  findAndPickSheetBarcodeItemSubtitle = '%d/%d added';
  /**
    Title for the find and pick partial alert.
    Default is "Incomplete scan!"
    */
  findAndPickPartialAlertTitle = 'Incomplete scan!';
  /**
    Subtitle for the find and pick partial alert.
    Default is "Some codes are yet to be scanned. Proceed with submission anyway?"
    */
  findAndPickPartialAlertSubtitle = 'Some codes are yet to be scanned. Proceed with submission anyway?';
  /**
    Cancel button text for the find and pick partial alert.
    Default is "Cancel"
    */
  findAndPickPartialAlertCancelButton = 'Cancel';
  /**
    Submit button text for the find and pick partial alert.
    Default is "Yes, Submit"
    */
  findAndPickPartialAlertSubmitButton = 'Yes, Submit';
  /**
    Title for the find and pick complete alert.
    Default is "All codes found!"
    */
  findAndPickCompleteAlertTitle = 'All codes found!';
  /**
    Subtitle for the find and pick complete alert.
    Default is "You have successfully scanned all codes. Proceed with submission?"
    */
  findAndPickCompleteAlertSubtitle = 'You have successfully scanned all codes. Proceed with submission?';
  /**
    Cancel button text for the find and pick complete alert.
    Default is "Dismiss"
    */
  findAndPickCompleteAlertCancelButton = 'Dismiss';
  /**
    Submit button text for the find and pick complete alert.
    Default is "Submit"
    */
  findAndPickCompleteAlertSubmitButton = 'Submit';
  /**
    Accessibility description for the flash button.
    Default is "Toggle flash"
    */
  accessibilityDescriptionFlashButton = 'Toggle flash';
  /**
    Accessibility description for the zoom button.
    Default is "Toggle camera zoom"
    */
  accessibilityDescriptionZoomButton = 'Toggle camera zoom';
  /**
    Accessibility description for the flip camera button.
    Default is "Flip camera"
    */
  accessibilityDescriptionFlipCameraButton = 'Flip camera';

  /** @param source {@displayType `DeepPartial<BarcodeTextLocalization>`} */
  constructor(source = {}) {
    super();
    if (source.cameraPermissionEnableCameraTitle !== undefined) {
      this.cameraPermissionEnableCameraTitle = source.cameraPermissionEnableCameraTitle;
    }
    if (source.cameraPermissionEnableCameraExplanation !== undefined) {
      this.cameraPermissionEnableCameraExplanation = source.cameraPermissionEnableCameraExplanation;
    }
    if (source.cameraPermissionEnableCameraButton !== undefined) {
      this.cameraPermissionEnableCameraButton = source.cameraPermissionEnableCameraButton;
    }
    if (source.cameraPermissionCloseButton !== undefined) {
      this.cameraPermissionCloseButton = source.cameraPermissionCloseButton;
    }
    if (source.topBarCancelButton !== undefined) {
      this.topBarCancelButton = source.topBarCancelButton;
    }
    if (source.topBarTitle !== undefined) {
      this.topBarTitle = source.topBarTitle;
    }
    if (source.userGuidance !== undefined) {
      this.userGuidance = source.userGuidance;
    }
    if (source.barcodeInfoMappingLoadingMessage !== undefined) {
      this.barcodeInfoMappingLoadingMessage = source.barcodeInfoMappingLoadingMessage;
    }
    if (source.barcodeInfoMappingErrorStateTitle !== undefined) {
      this.barcodeInfoMappingErrorStateTitle = source.barcodeInfoMappingErrorStateTitle;
    }
    if (source.barcodeInfoMappingErrorStateSubtitle !== undefined) {
      this.barcodeInfoMappingErrorStateSubtitle = source.barcodeInfoMappingErrorStateSubtitle;
    }
    if (source.barcodeInfoMappingErrorStateRetryButton !== undefined) {
      this.barcodeInfoMappingErrorStateRetryButton = source.barcodeInfoMappingErrorStateRetryButton;
    }
    if (source.barcodeInfoMappingErrorStateCancelButton !== undefined) {
      this.barcodeInfoMappingErrorStateCancelButton = source.barcodeInfoMappingErrorStateCancelButton;
    }
    if (source.manualCountEditDialogTitle !== undefined) {
      this.manualCountEditDialogTitle = source.manualCountEditDialogTitle;
    }
    if (source.manualCountEditDialogInfo !== undefined) {
      this.manualCountEditDialogInfo = source.manualCountEditDialogInfo;
    }
    if (source.manualCountEditDialogUpdateButton !== undefined) {
      this.manualCountEditDialogUpdateButton = source.manualCountEditDialogUpdateButton;
    }
    if (source.manualCountEditDialogCancelButton !== undefined) {
      this.manualCountEditDialogCancelButton = source.manualCountEditDialogCancelButton;
    }
    if (source.sheetResetButton !== undefined) {
      this.sheetResetButton = source.sheetResetButton;
    }
    if (source.sheetClearAllButton !== undefined) {
      this.sheetClearAllButton = source.sheetClearAllButton;
    }
    if (source.sheetSubmitButton !== undefined) {
      this.sheetSubmitButton = source.sheetSubmitButton;
    }
    if (source.sheetStartScanningButton !== undefined) {
      this.sheetStartScanningButton = source.sheetStartScanningButton;
    }
    if (source.sheetPlaceholderTitle !== undefined) {
      this.sheetPlaceholderTitle = source.sheetPlaceholderTitle;
    }
    if (source.sheetPlaceholderSubtitle !== undefined) {
      this.sheetPlaceholderSubtitle = source.sheetPlaceholderSubtitle;
    }
    if (source.singleModeConfirmationCancelButton !== undefined) {
      this.singleModeConfirmationCancelButton = source.singleModeConfirmationCancelButton;
    }
    if (source.multipleModeSheetTitle !== undefined) {
      this.multipleModeSheetTitle = source.multipleModeSheetTitle;
    }
    if (source.findAndPickSheetTitle !== undefined) {
      this.findAndPickSheetTitle = source.findAndPickSheetTitle;
    }
    if (source.findAndPickSheetBarcodeItemSubtitle !== undefined) {
      this.findAndPickSheetBarcodeItemSubtitle = source.findAndPickSheetBarcodeItemSubtitle;
    }
    if (source.findAndPickPartialAlertTitle !== undefined) {
      this.findAndPickPartialAlertTitle = source.findAndPickPartialAlertTitle;
    }
    if (source.findAndPickPartialAlertSubtitle !== undefined) {
      this.findAndPickPartialAlertSubtitle = source.findAndPickPartialAlertSubtitle;
    }
    if (source.findAndPickPartialAlertCancelButton !== undefined) {
      this.findAndPickPartialAlertCancelButton = source.findAndPickPartialAlertCancelButton;
    }
    if (source.findAndPickPartialAlertSubmitButton !== undefined) {
      this.findAndPickPartialAlertSubmitButton = source.findAndPickPartialAlertSubmitButton;
    }
    if (source.findAndPickCompleteAlertTitle !== undefined) {
      this.findAndPickCompleteAlertTitle = source.findAndPickCompleteAlertTitle;
    }
    if (source.findAndPickCompleteAlertSubtitle !== undefined) {
      this.findAndPickCompleteAlertSubtitle = source.findAndPickCompleteAlertSubtitle;
    }
    if (source.findAndPickCompleteAlertCancelButton !== undefined) {
      this.findAndPickCompleteAlertCancelButton = source.findAndPickCompleteAlertCancelButton;
    }
    if (source.findAndPickCompleteAlertSubmitButton !== undefined) {
      this.findAndPickCompleteAlertSubmitButton = source.findAndPickCompleteAlertSubmitButton;
    }
    if (source.accessibilityDescriptionFlashButton !== undefined) {
      this.accessibilityDescriptionFlashButton = source.accessibilityDescriptionFlashButton;
    }
    if (source.accessibilityDescriptionZoomButton !== undefined) {
      this.accessibilityDescriptionZoomButton = source.accessibilityDescriptionZoomButton;
    }
    if (source.accessibilityDescriptionFlipCameraButton !== undefined) {
      this.accessibilityDescriptionFlipCameraButton = source.accessibilityDescriptionFlipCameraButton;
    }
  }
}
