// react-native-scanbot-barcode-scanner-sdk/lib/module/ui_v2/common/Common.js
import { PartiallyConstructible } from '../../utils';

/**
Value palette describing the colors of the scanner screens.
*/
export class Palette extends PartiallyConstructible {
  /**
    The primary color used for enabled elements.
    Default is "#C8193C"
    */
  sbColorPrimary = '#C8193C';
  /**
    The primary color used for disabled elements.
    Default is "#F5F5F5"
    */
  sbColorPrimaryDisabled = '#F5F5F5';
  /**
    The color used to convey a negative meaning.
    Default is "#FF3737"
    */
  sbColorNegative = '#FF3737';
  /**
    The color used to convey a positive meaning.
    Default is "#4EFFB4"
    */
  sbColorPositive = '#4EFFB4';
  /**
    The color used to convey warnings.
    Default is "#FFCE5C"
    */
  sbColorWarning = '#FFCE5C';
  /**
    The secondary color used for enabled elements.
    Default is "#FFEDEE"
    */
  sbColorSecondary = '#FFEDEE';
  /**
    The secondary color used for disabled elements.
    Default is "#F5F5F5"
    */
  sbColorSecondaryDisabled = '#F5F5F5';
  /**
    The color used for displaying elements on top of the primary color.
    Default is "#FFFFFF"
    */
  sbColorOnPrimary = '#FFFFFF';
  /**
    The color used for displaying elements on top of the secondary color.
    Default is "#C8193C"
    */
  sbColorOnSecondary = '#C8193C';
  /**
    The color used for surfaces.
    Default is "#FFFFFF"
    */
  sbColorSurface = '#FFFFFF';
  /**
    The color used for outlines.
    Default is "#EFEFEF"
    */
  sbColorOutline = '#EFEFEF';
  /**
    The alternative color used on top of surfaces.
    Default is "#707070"
    */
  sbColorOnSurfaceVariant = '#707070';
  /**
    The color used on top of surfaces.
    Default is "#000000"
    */
  sbColorOnSurface = '#000000';
  /**
    The color with a low alpha (transparency) value used for certain surfaces.
    Default is "#00000026"
    */
  sbColorSurfaceLow = '#00000026';
  /**
    The color with a high alpha (transparency) value used for certain surfaces.
    Default is "#0000007A"
    */
  sbColorSurfaceHigh = '#0000007A';
  /**
    The color with a very high alpha (transparency) value used to fill certain overlays.
    Default is "#000000A3"
    */
  sbColorModalOverlay = '#000000A3';

  /** @param source {@displayType `DeepPartial<Palette>`} */
  constructor(source = {}) {
    super();
    if (source.sbColorPrimary !== undefined) {
      this.sbColorPrimary = source.sbColorPrimary;
    }
    if (source.sbColorPrimaryDisabled !== undefined) {
      this.sbColorPrimaryDisabled = source.sbColorPrimaryDisabled;
    }
    if (source.sbColorNegative !== undefined) {
      this.sbColorNegative = source.sbColorNegative;
    }
    if (source.sbColorPositive !== undefined) {
      this.sbColorPositive = source.sbColorPositive;
    }
    if (source.sbColorWarning !== undefined) {
      this.sbColorWarning = source.sbColorWarning;
    }
    if (source.sbColorSecondary !== undefined) {
      this.sbColorSecondary = source.sbColorSecondary;
    }
    if (source.sbColorSecondaryDisabled !== undefined) {
      this.sbColorSecondaryDisabled = source.sbColorSecondaryDisabled;
    }
    if (source.sbColorOnPrimary !== undefined) {
      this.sbColorOnPrimary = source.sbColorOnPrimary;
    }
    if (source.sbColorOnSecondary !== undefined) {
      this.sbColorOnSecondary = source.sbColorOnSecondary;
    }
    if (source.sbColorSurface !== undefined) {
      this.sbColorSurface = source.sbColorSurface;
    }
    if (source.sbColorOutline !== undefined) {
      this.sbColorOutline = source.sbColorOutline;
    }
    if (source.sbColorOnSurfaceVariant !== undefined) {
      this.sbColorOnSurfaceVariant = source.sbColorOnSurfaceVariant;
    }
    if (source.sbColorOnSurface !== undefined) {
      this.sbColorOnSurface = source.sbColorOnSurface;
    }
    if (source.sbColorSurfaceLow !== undefined) {
      this.sbColorSurfaceLow = source.sbColorSurfaceLow;
    }
    if (source.sbColorSurfaceHigh !== undefined) {
      this.sbColorSurfaceHigh = source.sbColorSurfaceHigh;
    }
    if (source.sbColorModalOverlay !== undefined) {
      this.sbColorModalOverlay = source.sbColorModalOverlay;
    }
  }
}

/**
Configuration of the text field appearance.
*/
export class StyledText extends PartiallyConstructible {
  /**
    Determines whether the text field is visible or not.
    Default is true
    */
  visible = true;
  /**
    The value of the text field.
    Default is ""
    */
  text = '';
  /**
    The text color.
    Default is "#FFFFFF"
    */
  color = '#FFFFFF';
  /**
    Determines whether to enable drop shadows for the text.
    Default is false
    */
  useShadow = false;

  /** @param source {@displayType `DeepPartial<StyledText>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.text !== undefined) {
      this.text = source.text;
    }
    if (source.color !== undefined) {
      this.color = source.color;
    }
    if (source.useShadow !== undefined) {
      this.useShadow = source.useShadow;
    }
  }
}

/**
Configuration of the icon appearance.
*/
export class IconStyle extends PartiallyConstructible {
  /**
    Determines whether the icon is visible or not.
    Default is true
    */
  visible = true;
  /**
    The icon color.
    Default is "#FFFFFF"
    */
  color = '#FFFFFF';

  /** @param source {@displayType `DeepPartial<IconStyle>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.color !== undefined) {
      this.color = source.color;
    }
  }
}

/**
Configuration of the icon appearance on a button.
*/
export class IconButton extends PartiallyConstructible {
  /**
    Determines whether the icon is visible on the button.
    Default is true
    */
  visible = true;
  /**
    The icon color.
    Default is "#FFFFFF"
    */
  color = '#FFFFFF';
  /**
    The text to be read when the button is selected through accessibility mode.
    Default is ""
    */
  accessibilityDescription = '';

  /** @param source {@displayType `DeepPartial<IconButton>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.color !== undefined) {
      this.color = source.color;
    }
    if (source.accessibilityDescription !== undefined) {
      this.accessibilityDescription = source.accessibilityDescription;
    }
  }
}

/**
Configuration of the polygon appearance.
*/
export class PolygonStyle extends PartiallyConstructible {
  /**
    The color of the polygon outline.
    Default is "#FFFFFFFF"
    */
  strokeColor = '#FFFFFFFF';
  /**
    The fill color of the polygon.
    Default is "#FFFFFF30"
    */
  fillColor = '#FFFFFF30';
  /**
    The width of the polygon outline in dp.
    Default is 2.0
    */
  strokeWidth = 2.0;
  /**
    The corner radius of the polygon in dp.
    Default is 0.0
    */
  cornerRadius = 0.0;

  /** @param source {@displayType `DeepPartial<PolygonStyle>`} */
  constructor(source = {}) {
    super();
    if (source.strokeColor !== undefined) {
      this.strokeColor = source.strokeColor;
    }
    if (source.fillColor !== undefined) {
      this.fillColor = source.fillColor;
    }
    if (source.strokeWidth !== undefined) {
      this.strokeWidth = source.strokeWidth;
    }
    if (source.cornerRadius !== undefined) {
      this.cornerRadius = source.cornerRadius;
    }
  }
}

/**
Configuration of the background appearance for buttons and hints.
*/
export class BackgroundStyle extends PartiallyConstructible {
  /**
    The color of the outline.
    Default is "#FFFFFFFF"
    */
  strokeColor = '#FFFFFFFF';
  /**
    The fill color.
    Default is "#FFFFFF30"
    */
  fillColor = '#FFFFFF30';
  /**
    The width of the outline in dp.
    Default is 2.0
    */
  strokeWidth = 2.0;

  /** @param source {@displayType `DeepPartial<BackgroundStyle>`} */
  constructor(source = {}) {
    super();
    if (source.strokeColor !== undefined) {
      this.strokeColor = source.strokeColor;
    }
    if (source.fillColor !== undefined) {
      this.fillColor = source.fillColor;
    }
    if (source.strokeWidth !== undefined) {
      this.strokeWidth = source.strokeWidth;
    }
  }
}

/**
Configuration of the appearance for foreground elements (e.g. text and/or icons, etc).
*/
export class ForegroundStyle extends PartiallyConstructible {
  /**
    Determines whether the icon is visible or not.
    Default is true
    */
  iconVisible = true;
  /**
    The color used for foreground elements.
    Default is "#FFFFFF"
    */
  color = '#FFFFFF';
  /**
    Determines whether to use drop shadows for foreground elements.
    Default is false
    */
  useShadow = false;

  /** @param source {@displayType `DeepPartial<ForegroundStyle>`} */
  constructor(source = {}) {
    super();
    if (source.iconVisible !== undefined) {
      this.iconVisible = source.iconVisible;
    }
    if (source.color !== undefined) {
      this.color = source.color;
    }
    if (source.useShadow !== undefined) {
      this.useShadow = source.useShadow;
    }
  }
}

/**
Configuration of the finder's aspect ratio.
*/
export class AspectRatio extends PartiallyConstructible {
  /**
    The width component of the aspect ratio.
    Default is 1.0
    */
  width = 1.0;
  /**
    The height component of the aspect ratio.
    Default is 1.0
    */
  height = 1.0;

  /** @param source {@displayType `DeepPartial<AspectRatio>`} */
  constructor(source = {}) {
    super();
    if (source.width !== undefined) {
      this.width = source.width;
    }
    if (source.height !== undefined) {
      this.height = source.height;
    }
  }
}

/**
Configuration of the badge.
*/
export class BadgeStyle extends PartiallyConstructible {
  /**
    Determines whether the badge is visible or not.
    Default is true
    */
  visible = true;
  /**
    Configuration of the background appearance for the badge.
    */
  background = new BackgroundStyle({});
  /**
    The color of the badge's foreground (icon, text).
    Default is "?sbColorOnSurface"
    */
  foregroundColor = '?sbColorOnSurface';

  /** @param source {@displayType `DeepPartial<BadgeStyle>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.background !== undefined) {
      this.background = new BackgroundStyle(source.background);
    }
    if (source.foregroundColor !== undefined) {
      this.foregroundColor = source.foregroundColor;
    }
  }
}

/**
Configuration of the round button.
*/
export class RoundButton extends PartiallyConstructible {
  /**
    Determines whether the button is visible or not.
    Default is true
    */
  visible = true;
  /**
    The text to be read when the button is selected through accessibility mode.
    Default is ""
    */
  accessibilityDescription = '';
  /**
    The color of the button's background.
    Default is "#0000007A"
    */
  backgroundColor = '#0000007A';
  /**
    The color of the button's foreground (icon, text).
    Default is "#FFFFFF"
    */
  foregroundColor = '#FFFFFF';
  /**
    The color of the button's background when the button is active (selected, toggled).
    Default is "#FFCE5C"
    */
  activeBackgroundColor = '#FFCE5C';
  /**
    The color of the button's foreground (icon, text) when the button is active (selected, toggled).
    Default is "#1C1B1F"
    */
  activeForegroundColor = '#1C1B1F';

  /** @param source {@displayType `DeepPartial<RoundButton>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.accessibilityDescription !== undefined) {
      this.accessibilityDescription = source.accessibilityDescription;
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.foregroundColor !== undefined) {
      this.foregroundColor = source.foregroundColor;
    }
    if (source.activeBackgroundColor !== undefined) {
      this.activeBackgroundColor = source.activeBackgroundColor;
    }
    if (source.activeForegroundColor !== undefined) {
      this.activeForegroundColor = source.activeForegroundColor;
    }
  }
}

/**
Configuration of the button with a badge.
*/
export class BadgedButton extends PartiallyConstructible {
  /**
    The color of the badge's background.
    Default is "#FFFFFF"
    */
  badgeBackgroundColor = '#FFFFFF';
  /**
    The color of the badge's foreground (icon, text).
    Default is "#C8193C"
    */
  badgeForegroundColor = '#C8193C';
  /**
    Determines whether the button is visible or not.
    Default is true
    */
  visible = true;
  /**
    The color of the button's background.
    Default is "#0000007A"
    */
  backgroundColor = '#0000007A';
  /**
    The color of the button's foreground (icon, text).
    Default is "#FFFFFF"
    */
  foregroundColor = '#FFFFFF';
  /**
    The color of the button's background when the button is active (selected, toggled).
    Default is "#FFCE5C"
    */
  activeBackgroundColor = '#FFCE5C';
  /**
    The color of the button's foreground (icon, text) when the button is active (selected, toggled).
    Default is "#1C1B1F"
    */
  activeForegroundColor = '#1C1B1F';

  /** @param source {@displayType `DeepPartial<BadgedButton>`} */
  constructor(source = {}) {
    super();
    if (source.badgeBackgroundColor !== undefined) {
      this.badgeBackgroundColor = source.badgeBackgroundColor;
    }
    if (source.badgeForegroundColor !== undefined) {
      this.badgeForegroundColor = source.badgeForegroundColor;
    }
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.foregroundColor !== undefined) {
      this.foregroundColor = source.foregroundColor;
    }
    if (source.activeBackgroundColor !== undefined) {
      this.activeBackgroundColor = source.activeBackgroundColor;
    }
    if (source.activeForegroundColor !== undefined) {
      this.activeForegroundColor = source.activeForegroundColor;
    }
  }
}

/**
Configuration of the button.
*/
export class ButtonConfiguration extends PartiallyConstructible {
  /**
    Determines whether the button is visible or not.
    Default is true
    */
  visible = true;
  /**
    The text to be displayed on the button.
    Default is ""
    */
  text = '';
  /**
    The text to be read when the button is selected through accessibility mode.
    Default is ""
    */
  accessibilityDescription = '';
  /**
    Configuration of the background appearance for the button.
    */
  background = new BackgroundStyle({});
  /**
    Configuration of the appearance for foreground elements (e.g. text and/or icons, etc) of the button.
    */
  foreground = new ForegroundStyle({});

  /** @param source {@displayType `DeepPartial<ButtonConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.text !== undefined) {
      this.text = source.text;
    }
    if (source.accessibilityDescription !== undefined) {
      this.accessibilityDescription = source.accessibilityDescription;
    }
    if (source.background !== undefined) {
      this.background = new BackgroundStyle(source.background);
    }
    if (source.foreground !== undefined) {
      this.foreground = new ForegroundStyle(source.foreground);
    }
  }
}

/**
Configuration for the popup menu items.
*/
export class PopupMenuItem extends PartiallyConstructible {
  /**
    The text to be displayed on the button.
    */
  title = new StyledText({});
  /**
    The text to be read when the button is selected through accessibility mode.
    Default is ""
    */
  accessibilityDescription = '';
  /**
    Configuration of the icon appearance.
    */
  icon = new IconStyle({});

  /** @param source {@displayType `DeepPartial<PopupMenuItem>`} */
  constructor(source = {}) {
    super();
    if (source.title !== undefined) {
      this.title = new StyledText(source.title);
    }
    if (source.accessibilityDescription !== undefined) {
      this.accessibilityDescription = source.accessibilityDescription;
    }
    if (source.icon !== undefined) {
      this.icon = new IconStyle(source.icon);
    }
  }
}

/**
Configuration of the button located on a bar.
*/
export class BarButtonConfiguration extends PartiallyConstructible {
  /**
    Determines whether the button is visible or not.
    Default is true
    */
  visible = true;
  /**
    The text to be displayed on the button.
    */
  title = new StyledText({});
  /**
    The text to be read when the button is selected through accessibility mode.
    Default is ""
    */
  accessibilityDescription = '';
  /**
    Configuration of the background appearance for the button.
    */
  background = new BackgroundStyle({});
  /**
    Configuration of the icon appearance.
    */
  icon = new IconStyle({});

  /** @param source {@displayType `DeepPartial<BarButtonConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.title !== undefined) {
      this.title = new StyledText(source.title);
    }
    if (source.accessibilityDescription !== undefined) {
      this.accessibilityDescription = source.accessibilityDescription;
    }
    if (source.background !== undefined) {
      this.background = new BackgroundStyle(source.background);
    }
    if (source.icon !== undefined) {
      this.icon = new IconStyle(source.icon);
    }
  }
}

/**
Configure the orientation of the interface.

- `NONE`:
   Do not restrict the interface's orientation.
- `PORTRAIT`:
   Lock the orientation to portrait.
- `LANDSCAPE`:
   Lock the orientation to landscape.
*/

/**
Configuration of the camera preview mode.

- `FIT_IN`:
   In this mode, the camera preview frames will be scaled to fit inside the layout view size - the full preview frame content will be visible, but unused edges might appear in the preview layout.
- `FILL_IN`:
   In this mode, the camera preview frames fill the entire layout view - the preview frames may contain additional content at the edges that are not visible in the preview layout.
*/

/**
Determines the successful detection sound.

- `MODERN_BEEP`:
   A modern beep sound.
- `CLASSIC_BEEP`:
   The old, classic beep sound.
*/

/**
Configuration for the sound.
*/
export class Sound extends PartiallyConstructible {
  /**
    Determine whether the beep sound should be enabled or not when a barcode is detected.
    Default is true
    */
  successBeepEnabled = true;
  /**
    Determines the successful detection sound.
    Default is MODERN_BEEP
    */
  soundType = 'MODERN_BEEP';

  /** @param source {@displayType `DeepPartial<Sound>`} */
  constructor(source = {}) {
    super();
    if (source.successBeepEnabled !== undefined) {
      this.successBeepEnabled = source.successBeepEnabled;
    }
    if (source.soundType !== undefined) {
      this.soundType = source.soundType;
    }
  }
}

/**
Configure the vibration.
*/
export class Vibration extends PartiallyConstructible {
  /**
    Determine whether vibration should be enabled or not when a barcode is detected.
    Default is false
    */
  enabled = false;

  /** @param source {@displayType `DeepPartial<Vibration>`} */
  constructor(source = {}) {
    super();
    if (source.enabled !== undefined) {
      this.enabled = source.enabled;
    }
  }
}

/**
Configuration of timeouts.
*/
export class Timeouts extends PartiallyConstructible {
  /**
    Sets the length of time, in milliseconds, when the scanner should auto close. Default is 0 (disabled).
    Default is 0
    */
  autoCancelTimeout = 0;
  /**
    Sets the length of time, in milliseconds, that the first scan will be delayed by. Default is 0 (disabled).
    Default is 0
    */
  initialScanDelay = 0;

  /** @param source {@displayType `DeepPartial<Timeouts>`} */
  constructor(source = {}) {
    super();
    if (source.autoCancelTimeout !== undefined) {
      this.autoCancelTimeout = source.autoCancelTimeout;
    }
    if (source.initialScanDelay !== undefined) {
      this.initialScanDelay = source.initialScanDelay;
    }
  }
}
