// react-native-scanbot-barcode-scanner-sdk/lib/module/ui_v2/common/UserGuidanceConfiguration.js
import { BackgroundStyle } from '../common/Common';
import { PartiallyConstructible } from '../../utils';
import { IconStyle } from '../common/Common';
import { StyledText } from '../common/Common';

/**
Configuration of the hint guiding users through the scanning process.
*/
export class UserGuidanceConfiguration extends PartiallyConstructible {
  /**
    Determines whether the user guidance is visible or not.
    Default is true
    */
  visible = true;
  /**
    The title of the user guidance.
    */
  title = new StyledText({
    color: '?sbColorOnPrimary'
  });
  /**
    The background style used for the user guidance.
    */
  background = new BackgroundStyle({
    strokeColor: '#00000000',
    fillColor: '?sbColorSurfaceLow'
  });

  /** @param source {@displayType `DeepPartial<UserGuidanceConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.title !== undefined) {
      this.title = new StyledText(source.title);
    }
    if (source.background !== undefined) {
      this.background = new BackgroundStyle(source.background);
    }
  }
}

/**
Configuration of the hint (containing an icon) guiding users through the scanning process.
*/
export class IconUserGuidanceConfiguration extends PartiallyConstructible {
  /**
    Determines whether the user guidance is visible or not.
    Default is true
    */
  visible = true;
  /**
    Configuration of the icon appearance.
    */
  icon = new IconStyle({
    color: '?sbColorOnPrimary'
  });
  /**
    The title of the user guidance.
    */
  title = new StyledText({
    color: '?sbColorOnPrimary'
  });
  /**
    Configuration of the background appearance for the user guidance hints.
    */
  background = new BackgroundStyle({
    fillColor: '?sbColorSurfaceLow'
  });

  /** @param source {@displayType `DeepPartial<IconUserGuidanceConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
    if (source.icon !== undefined) {
      this.icon = new IconStyle(source.icon);
    }
    if (source.title !== undefined) {
      this.title = new StyledText(source.title);
    }
    if (source.background !== undefined) {
      this.background = new BackgroundStyle(source.background);
    }
  }
}
